modules/kafka/kafka-dest-worker.c

```diff
@@ -61,16 +62,38 @@ _publish_message(KafkaDestWorker *self, LogMessage *msg)
 {
   KafkaDestDriver *owner = (KafkaDestDriver *) self->super.owner;
   int block_flag = _is_poller_thread(self) ? 0 : RD_KAFKA_MSG_F_BLOCK;
+  rd_kafka_topic_t *topic;
+  //GString *topic_name_new = g_string_sized_new(128);
+  GString *topic_name_new = scratch_buffers_alloc();
```

Comment:
I think there could be a separate function to get the necessary topic name, instead of the `if-else`s in this function. From this code it is hard to determine what is used in the end and why.

Reply:
Given that there are multiple variables inside the function which get updated, how do I do this?

Reply:
There is no need for the `is_fallback_topic` variable, you can just write a warning when it is used.
Something like this:

```
static void
_calculate_topic_name_from_template(KafkaDestWorker *self, LogMessage *msg, GString *topic_name)
{
  // Throw a warning message if fallback topic is used
  // Use g_string_assign()
  ...
}

static void
_get_literal_topic_name(KafkaDestWorker *self, LogMessage *msg, GString *topic_name)
{
  // Use g_string_assign()
  ...
}

static void
_calculate_topic_name(KafkaDestWorker *self, LogMessage *msg, GString *topic_name)
{
  if (kafka_dd_is_topic_name_a_template(owner))
    _calculate_topic_name_from_template(self, msg, topic_name);

  _get_literal_topic_name(self, msg, topic_name);
}

static gboolean
_publish_message(KafkaDestWorker *self, LogMessage *msg)
{
...
  GString *topic_name = scratch_buffers_alloc_and_mark(&marker);
  _calculate_topic_name(self, msg, topic_name);

  if (rd_kafka_produce(topic,
                       RD_KAFKA_PARTITION_UA,
                       RD_KAFKA_MSG_F_FREE | block_flag,
                       self->message->str, self->message->len,
                       self->key->len ? self->key->str : NULL, self->key->len,
                       log_msg_ref(msg)) == -1)
    {
      msg_error("kafka: failed to publish message",
                evt_tag_str("topic", topic_name->str),
                evt_tag_str("error", rd_kafka_err2str(rd_kafka_last_error())),
                evt_tag_str("driver", owner->super.super.super.id),
                log_pipe_location_tag(&owner->super.super.super.super));

      scratch_buffers_reclaim_marked(marker);
      return FALSE;
    }

  msg_debug("kafka: message published",
            evt_tag_str("topic", topic_name->str),
            evt_tag_str("key", self->key->len ? self->key->str : "NULL"),
            evt_tag_str("message", self->message->str),
            evt_tag_str("driver", owner->super.super.super.id),
            log_pipe_location_tag(&owner->super.super.super.super));

  /* we passed the allocated buffers to rdkafka, which will eventually free them */
  g_string_steal(self->message);
  scratch_buffers_reclaim_marked(marker);
  return TRUE;
}
```

If the `topic_name_buffer` is implemented, you will no longer have to pass a `GString *` to `_calculate_topic_name()`, but you can use the worker's buffer.